fog/ledger/enclave/impl/src/oblivious_utils.rs

```diff
@@ -0,0 +1,454 @@
+                false,
+                "Shard result is unspent but overwrites spent client!",
+            ),
+        ];
+
+        for (
+            client_key_image,
+            client_spent_at,
+            client_timestamp,
+            client_timestamp_result_code,
+            client_key_image_result_code,
+            shard_key_image,
+            shard_spent_at,
+            shard_timestamp,
+            shard_timestamp_result_code,
+            shard_key_image_result_code,
+            expected_result,
+            panic_message,
+        ) in test_cases.into_iter()
+        {
+            let client_result = KeyImageResult {
+                key_image: client_key_image.into(),
+                spent_at: client_spent_at,
+                timestamp: client_timestamp,
+                timestamp_result_code: client_timestamp_result_code as u32,
+                key_image_result_code: client_key_image_result_code as u32,
+            };
+            let shard_result = KeyImageResult {
+                key_image: shard_key_image.into(),
+                spent_at: shard_spent_at,
+                timestamp: shard_timestamp,
+                timestamp_result_code: shard_timestamp_result_code as u32,
+                key_image_result_code: shard_key_image_result_code as u32,
+            };
+            let result: bool =
+                should_overwrite_key_image_search_result(&client_result, &shard_result).into();
+            assert_eq!(result, expected_result, "{}", panic_message);
+        }
+    }
+
+    fn collate(
+        client_queries: Vec<KeyImageQuery>,
+        shard_key_image_search_results: Vec<KeyImageResult>,
+    ) -> Vec<KeyImageResult> {
+        let queries_count = client_queries.len();
+        let result =
+            collate_shard_key_image_search_results(client_queries, shard_key_image_search_results);
+
+        let results_count = result.len();
+        assert_eq!(
+            results_count, queries_count,
+            "{}",
+            "results length does not match number of queries"
+        );
+
+        result
+    }
+
+    #[test]
+    fn collation_tests() {
+        // All results available, no dupes
+        let client_queries = vec![KeyImageQuery {
+            key_image: 1.into(),
+            start_block: 0,
+        }];
+
+        let shard_results = vec![KeyImageResult {
+            key_image: 1.into(),
+            spent_at: 1,
+            timestamp: 123,
+            timestamp_result_code: TimestampResultCode::TimestampFound as u32,
+            key_image_result_code: KeyImageResultCode::Spent as u32,
+        }];
+
+        let result = collate(client_queries, shard_results);
+
+        let all_images_found = result.iter().all(|key_image_result| {
+            key_image_result.key_image_result_code == KeyImageResultCode::Spent as u32
+        });
+
+        assert!(
+            all_images_found,
+            "{}",
+            "all images are present but some were not collated"
+        );
```

Comment:
:wrench: A couple things on these _tests_:
- It looks like after each assert it's a new test case, suggest making each one it's own function
- Be more explicit on the expected, it looks like only the counts and the result code is compared, but the entire image should be copied over.
- There is no test for mixed results codes, i.e. image 1 has spent, 2 has error

Suggest something like the following:
- it makes it easier to have more than one item per Vec
- it makes it more explicit what the expected is
- The duplicate entries are more explicit.
- Chose a `TimestampResultCode` that differs from the default (only needs to be done in one test case
```suggestion
    #[test]
    fn all_available() {
        let range = 1..=3;
        let client_queries = range
            .clone()
            .map(|key_image| KeyImageQuery {
                key_image: key_image.into(),
                start_block: 0,
            })
            .collect::<Vec<_>>();
        let mut shard_results = range
            .map(|key_image| KeyImageResult {
                key_image: key_image.into(),
                spent_at: key_image + 1,
                timestamp: key_image + 10,
                timestamp_result_code: TimestampResultCode::WatcherBehind as u32,
                key_image_result_code: KeyImageResultCode::Spent as u32,
            })
            .collect::<Vec<_>>();

        let mut results = collate_shard_key_image_search_results(client_queries, &shard_results);
        results.sort_by_key(|r| r.key_image);
        shard_results.sort_by_key(|r| r.key_image);
        assert_eq!(results, shard_results);
    }

    #[test]
    fn duplicate_shard_results_returns_one_result() {
        let client_queries = vec![KeyImageQuery {
            key_image: 1.into(),
            start_block: 0,
        }];
        let key_image_result = KeyImageResult {
            key_image: 1.into(),
            spent_at: 2,
            timestamp: 3,
            timestamp_result_code: TimestampResultCode::WatcherBehind as u32,
            key_image_result_code: KeyImageResultCode::Spent as u32,
        };
        let shard_results = vec![key_image_result.clone(), key_image_result.clone()];
        let mut results = collate_shard_key_image_search_results(client_queries, &shard_results);
        results.sort_by_key(|r| r.key_image);
        assert_eq!(results, vec![key_image_result]);
    }
```